test_conformance/subgroups/test_subgroup_ballot.cpp

```diff
@@ -0,0 +1,1089 @@
+    "    out[gid] = value;\n"
+    "}\n";
+
+static const char *ballot_find_msb_source =
+    "__kernel void test_sub_group_ballot_find_msb(const __global Type *in, "
+    "__global int4 *xy, __global Type *out)\n"
+    "{\n"
+    "    int gid = get_global_id(0);\n"
+    "    XY(xy,gid);\n"
+    "    Type x = in[gid];\n"
+    "    uint4 value = (uint4)(0,0,0,0);"
+    "    value = (uint4)(sub_group_ballot_find_msb(x),0,0,0);"
+    "    out[gid] = value ;"
+    "}\n";
+
+static const char *get_subgroup_ge_mask_source =
+    "__kernel void test_get_sub_group_ge_mask(const __global Type *in, "
+    "__global int4 *xy, __global Type *out)\n"
+    "{\n"
+    "    int gid = get_global_id(0);\n"
+    "    XY(xy,gid);\n"
+    "    xy[gid].z = get_max_sub_group_size();\n"
+    "    Type x = in[gid];\n"
+    "    uint4 mask = get_sub_group_ge_mask();"
+    "    out[gid] = mask;\n"
+    "}\n";
+
+static const char *get_subgroup_gt_mask_source =
+    "__kernel void test_get_sub_group_gt_mask(const __global Type *in, "
+    "__global int4 *xy, __global Type *out)\n"
+    "{\n"
+    "    int gid = get_global_id(0);\n"
+    "    XY(xy,gid);\n"
+    "    xy[gid].z = get_max_sub_group_size();\n"
+    "    Type x = in[gid];\n"
+    "    uint4 mask = get_sub_group_gt_mask();"
+    "    out[gid] = mask;\n"
+    "}\n";
+
+static const char *get_subgroup_le_mask_source =
+    "__kernel void test_get_sub_group_le_mask(const __global Type *in, "
+    "__global int4 *xy, __global Type *out)\n"
+    "{\n"
+    "    int gid = get_global_id(0);\n"
+    "    XY(xy,gid);\n"
+    "    xy[gid].z = get_max_sub_group_size();\n"
+    "    Type x = in[gid];\n"
+    "    uint4 mask = get_sub_group_le_mask();"
+    "    out[gid] = mask;\n"
+    "}\n";
+
+static const char *get_subgroup_lt_mask_source =
+    "__kernel void test_get_sub_group_lt_mask(const __global Type *in, "
+    "__global int4 *xy, __global Type *out)\n"
+    "{\n"
+    "    int gid = get_global_id(0);\n"
+    "    XY(xy,gid);\n"
+    "    xy[gid].z = get_max_sub_group_size();\n"
+    "    Type x = in[gid];\n"
+    "    uint4 mask = get_sub_group_lt_mask();"
+    "    out[gid] = mask;\n"
+    "}\n";
+
+static const char *get_subgroup_eq_mask_source =
+    "__kernel void test_get_sub_group_eq_mask(const __global Type *in, "
+    "__global int4 *xy, __global Type *out)\n"
+    "{\n"
+    "    int gid = get_global_id(0);\n"
+    "    XY(xy,gid);\n"
+    "    xy[gid].z = get_max_sub_group_size();\n"
+    "    Type x = in[gid];\n"
+    "    uint4 mask = get_sub_group_eq_mask();"
+    "    out[gid] = mask;\n"
+    "}\n";
+
+static const char *ballot_source =
+    "__kernel void test_sub_group_ballot(const __global Type *in, "
+    "__global int4 *xy, __global Type *out)\n"
+    "{\n"
+    "uint4 full_ballot = sub_group_ballot(1);\n"
+    "uint divergence_mask;\n"
+    "uint4 partial_ballot;\n"
+    "uint gid = get_global_id(0);"
+    "XY(xy,gid);\n"
+    "if (get_sub_group_local_id() & 1) {\n"
```

Comment:
It seems that the example I gave of divergence has been directly copied and pasting into the testing, here, without appropriate modifications.

The testing of non-uniform control flow should be similar (or identical) to sub_group_non_uniform_broadcast(), sub_group_elect() and in fact all of the extended subgroup functions that are sensitive to divergence of control flow within the subgroup.  Please update it to match, once the testing for sub_group_elect(), etc. correctly supports a maximum subgroup size of 128.

Reply:
This test is not a test of the sub_group_ballot() function.

It will pass even if the implementation of sub_group_ballot() unconditionally returns 0, always.

Reply:
Agreed.